File: packages/shared/src/schemas/posture.ts
import { z } from "zod";
import { UUIDSchema, DateTimeSchema, PaginationParamsSchema, PaginatedResponseSchema } from "./common";

export const CoverageSchema = z.record(z.object({
  totalTechniques: z.number().int().nonnegative(),
  coveredTechniques: z.number().int().nonnegative(),
  coverage: z.number().min(0).max(1)
}));
export type Coverage = z.infer<typeof CoverageSchema>;

export const PostureScoreSchema = z.object({
  id: UUIDSchema,
  episodeId: UUIDSchema,
  projectId: UUIDSchema,
  detectionRate: z.number().min(0).max(1),
  mttrSeconds: z.number().nonnegative(),
  coverage: CoverageSchema,
  overallScore: z.number().min(0).max(100),
  trend: z.enum(["improving", "stable", "declining"]).optional(),
  previousScore: z.number().min(0).max(100).optional(),
  calculatedAt: DateTimeSchema
});
export type PostureScore = z.infer<typeof PostureScoreSchema>;

export const PostureScoreTrendSchema = z.object({
  projectId: UUIDSchema,
  scores: z.array(z.object({
    episodeId: UUIDSchema,
    overallScore: z.number().min(0).max(100),
    detectionRate: z.number().min(0).max(1),
    mttrSeconds: z.number().nonnegative(),
    calculatedAt: DateTimeSchema
  })),
  trend: z.enum(["improving", "stable", "declining"]),
  improvementRate: z.number()
});
export type PostureScoreTrend = z.infer<typeof PostureScoreTrendSchema>;

export const PostureScoreListParamsSchema = PaginationParamsSchema.extend({
  projectId: UUIDSchema.optional(),
  episodeId: UUIDSchema.optional(),
  dateFrom: DateTimeSchema.optional(),
  dateTo: DateTimeSchema.optional()
});
export type PostureScoreListParams = z.infer<typeof PostureScoreListParamsSchema>;

export const PostureScoreListResponseSchema = PaginatedResponseSchema(PostureScoreSchema);
export type PostureScoreListResponse = z.infer<typeof PostureScoreListResponseSchema>;

export const PostureSummarySchema = z.object({
  projectId: UUIDSchema,
  currentScore: z.number().min(0).max(100),
  previousScore: z.number().min(0).max(100).optional(),
  trend: z.enum(["improving", "stable", "declining"]),
  detectionRate: z.number().min(0).max(1),
  mttrSeconds: z.number().nonnegative(),
  coverageByCategory: z.record(z.number().min(0).max(1)),
  totalEpisodes: z.number().int().nonnegative(),
  lastCalculatedAt: DateTimeSchema
});
export type PostureSummary = z.infer<typeof PostureSummarySchema>;

File: apps/api/app/schemas/posture.py
from datetime import datetime
from uuid import UUID
from typing import Optional
from pydantic import BaseModel, ConfigDict, Field

from app.schemas.common import PaginationParams, PaginatedResponse


class CoverageItem(BaseModel):
    total_techniques: int = 0
    covered_techniques: int = 0
    coverage: float = 0.0


class PostureScoreResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    episode_id: UUID
    project_id: UUID
    detection_rate: float
    mttr_seconds: int
    coverage: dict[str, CoverageItem]
    overall_score: float
    trend: Optional[str] = None
    previous_score: Optional[float] = None
    calculated_at: datetime


class PostureScoreTrendItem(BaseModel):
    episode_id: UUID
    overall_score: float
    detection_rate: float
    mttr_seconds: int
    calculated_at: datetime


class PostureScoreTrend(BaseModel):
    project_id: UUID
    scores: list[PostureScoreTrendItem]
    trend: str
    improvement_rate: float


class PostureScoreListParams(PaginationParams):
    project_id: Optional[UUID] = None
    episode_id: Optional[UUID] = None
    date_from: Optional[datetime] = None
    date_to: Optional[datetime] = None


class PostureScoreListResponse(PaginatedResponse[PostureScoreResponse]):
    pass


class PostureSummary(BaseModel):
    project_id: UUID
    current_score: float
    previous_score: Optional[float] = None
    trend: str
    detection_rate: float
    mttr_seconds: int
    coverage_by_category: dict[str, float]
    total_episodes: int
    last_calculated_at: datetime